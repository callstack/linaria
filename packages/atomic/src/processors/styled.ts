import type { SourceLocation } from '@babel/types';

import { debug } from '@linaria/logger';
import type { IProps } from '@linaria/react/processors/styled';
import StyledProcessor from '@linaria/react/processors/styled';
import { hasMeta } from '@linaria/tags';
import type { Rules, ValueCache } from '@linaria/tags';

import atomize from './helpers/atomize';

export default class AtomicStyledProcessor extends StyledProcessor {
  #classes: string | undefined;

  private get classes(): string {
    if (this.#classes) {
      return this.#classes;
    }

    throw new Error(
      'Styles are not extracted yet. Please call `extractRules` first.'
    );
  }

  public override extractRules(
    valueCache: ValueCache,
    cssText: string,
    loc?: SourceLocation | null
  ): Rules {
    const rules: Rules = {};

    const wrappedValue =
      typeof this.component === 'string'
        ? null
        : valueCache.get(this.component.node.name);

    const atomicRules = atomize(cssText, hasMeta(wrappedValue));
    atomicRules.forEach((rule) => {
      // eslint-disable-next-line no-param-reassign
      rules[rule.cssText] = {
        cssText: rule.cssText,
        start: loc?.start ?? null,
        className: this.className,
        displayName: this.displayName,
        atom: true,
      };

      debug(
        'evaluator:template-processor:extracted-atomic-rule',
        `\n${rule.cssText}`
      );
    });

    this.#classes = atomicRules
      // Some atomic rules produced (eg. keyframes) don't have class names, and they also don't need to appear in the object
      .filter((rule) => !!rule.className)
      .map((rule) => rule.className!)
      .join(' ');

    return rules;
  }

  protected override getProps(): IProps {
    const props = super.getProps();
    props.class = [this.classes, this.className].filter(Boolean).join(' ');
    props.atomic = true;
    return props;
  }

  protected override getVariableId(
    source: string,
    unit: string,
    precedingCss: string
  ): string {
    const id = this.getCustomVariableId(source, unit, precedingCss);
    if (id) {
      return id;
    }

    const context = this.getVariableContext(source, unit, precedingCss);
    // id is based on the slugified value
    return context.valueSlug;
  }
}
