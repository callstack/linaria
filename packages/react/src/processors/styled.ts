import { readFileSync } from 'fs';
import { dirname, join, sep } from 'path';

import type {
  CallExpression,
  Expression,
  ObjectExpression,
  SourceLocation,
  StringLiteral,
} from '@babel/types';
import { minimatch } from 'minimatch';
import html from 'react-html-attributes';

import type {
  Params,
  Rules,
  TailProcessorParams,
  ValueCache,
  WrappedNode,
} from '@linaria/tags';
import {
  buildSlug,
  hasMeta,
  TaggedTemplateProcessor,
  validateParams,
  ValueType,
  toValidCSSIdentifier,
} from '@linaria/tags';
import type { IVariableContext } from '@linaria/utils';
import { findPackageJSON, slugify } from '@linaria/utils';

const isNotNull = <T>(x: T | null): x is T => x !== null;

const allTagsSet = new Set([...html.elements.html, html.elements.svg]);

export interface IProps {
  atomic?: boolean;
  class?: string;
  name: string;
  propsAsIs: boolean;
  vars?: Record<string, Expression[]>;
}

const singleQuotedStringLiteral = (value: string): StringLiteral => ({
  type: 'StringLiteral',
  value,
  extra: {
    rawValue: value,
    raw: `'${value}'`,
  },
});

export default class StyledProcessor extends TaggedTemplateProcessor {
  public component: WrappedNode;

  #variableIdx = 0;

  #variablesCache = new Map<string, string>();

  constructor(params: Params, ...args: TailProcessorParams) {
    // Should have at least two params and the first one should be a callee.
    validateParams(
      params,
      ['callee', '*', '...'],
      TaggedTemplateProcessor.SKIP
    );

    validateParams(
      params,
      ['callee', ['call', 'member'], ['template', 'call']],
      'Invalid usage of `styled` tag'
    );

    const [tag, tagOp, template] = params;

    if (template[0] === 'call') {
      // It is already transformed styled-literal. Skip it.
      // eslint-disable-next-line @typescript-eslint/no-throw-literal
      throw TaggedTemplateProcessor.SKIP;
    }

    super([tag, template], ...args);

    let component: WrappedNode | undefined;
    if (tagOp[0] === 'call' && tagOp.length === 2) {
      const value = tagOp[1];
      if (value.kind === ValueType.FUNCTION) {
        component = 'FunctionalComponent';
      } else if (value.kind === ValueType.CONST) {
        component = typeof value.value === 'string' ? value.value : undefined;
      } else {
        if (value.importedFrom?.length) {
          const selfPkg = findPackageJSON('.', this.context.filename);

          // Check if at least one used identifier is a Linaria component.
          const isSomeMatched = value.importedFrom.some((importedFrom) => {
            const importedPkg = findPackageJSON(
              importedFrom,
              this.context.filename
            );

            if (importedPkg) {
              const packageJSON = JSON.parse(readFileSync(importedPkg, 'utf8'));
              let mask: string | undefined = packageJSON?.linaria?.components;
              if (importedPkg === selfPkg && mask === undefined) {
                // If mask is not specified for the local package, all components are treated as styled.
                mask = '**/*';
              }

              if (mask) {
                const packageDir = dirname(importedPkg);
                const normalizedMask = mask.replace(/\//g, sep);
                const fullMask = join(packageDir, normalizedMask);
                const fileWithComponent = require.resolve(importedFrom, {
                  paths: [dirname(this.context.filename!)],
                });

                return minimatch(fileWithComponent, fullMask);
              }
            }

            return false;
          });

          if (!isSomeMatched) {
            component = {
              node: value.ex,
              nonLinaria: true,
              source: value.source,
            };
          }
        }

        if (component === undefined) {
          component = {
            node: value.ex,
            source: value.source,
          };

          this.dependencies.push(value);
        }
      }
    }

    if (tagOp[0] === 'member') {
      [, component] = tagOp;
    }

    if (!component) {
      throw new Error('Invalid usage of `styled` tag');
    }

    this.component = component;
  }

  public override addInterpolation(
    node: Expression,
    precedingCss: string,
    source: string,
    unit = ''
  ): string {
    const id = this.getVariableId(source, unit, precedingCss);

    this.interpolations.push({
      id,
      node,
      source,
      unit,
    });

    return id;
  }

  public override doEvaltimeReplacement(): void {
    this.replacer(this.value, false);
  }

  public override doRuntimeReplacement(): void {
    const t = this.astService;

    const props = this.getProps();

    this.replacer(
      t.callExpression(this.tagExpression, [this.getTagComponentProps(props)]),
      true
    );
  }

  public override extractRules(
    valueCache: ValueCache,
    cssText: string,
    loc?: SourceLocation | null
  ): Rules {
    const rules: Rules = {};

    let selector = `.${this.className}`;

    // If `styled` wraps another component and not a primitive,
    // get its class name to create a more specific selector
    // it'll ensure that styles are overridden properly
    let value =
      typeof this.component === 'string' || this.component.nonLinaria
        ? null
        : valueCache.get(this.component.node.name);
    while (hasMeta(value)) {
      selector += `.${value.__linaria.className}`;
      value = value.__linaria.extends;
    }

    rules[selector] = {
      cssText,
      className: this.className,
      displayName: this.displayName,
      start: loc?.start ?? null,
    };

    return rules;
  }

  public override get asSelector(): string {
    return `.${this.className}`;
  }

  protected get tagExpressionArgument(): Expression {
    const t = this.astService;
    if (typeof this.component === 'string') {
      if (this.component === 'FunctionalComponent') {
        return t.arrowFunctionExpression([], t.blockStatement([]));
      }

      return singleQuotedStringLiteral(this.component);
    }

    return t.callExpression(t.identifier(this.component.node.name), []);
  }

  protected get tagExpression(): CallExpression {
    const t = this.astService;
    return t.callExpression(this.callee, [this.tagExpressionArgument]);
  }

  public override get value(): ObjectExpression {
    const t = this.astService;
    const extendsNode =
      typeof this.component === 'string' || this.component.nonLinaria
        ? null
        : this.component.node.name;

    return t.objectExpression([
      t.objectProperty(
        t.stringLiteral('displayName'),
        t.stringLiteral(this.displayName)
      ),
      t.objectProperty(
        t.stringLiteral('__linaria'),
        t.objectExpression([
          t.objectProperty(
            t.stringLiteral('className'),
            t.stringLiteral(this.className)
          ),
          t.objectProperty(
            t.stringLiteral('extends'),
            extendsNode
              ? t.callExpression(t.identifier(extendsNode), [])
              : t.nullLiteral()
          ),
        ])
      ),
    ]);
  }

  public override toString(): string {
    const res = (arg: string) => `${this.tagSourceCode()}(${arg})\`…\``;

    if (typeof this.component === 'string') {
      if (this.component === 'FunctionalComponent') {
        return res('() => {…}');
      }

      return res(`'${this.component}'`);
    }

    return res(this.component.source);
  }

  protected getCustomVariableId(
    source: string,
    unit: string,
    precedingCss: string
  ) {
    const context = this.getVariableContext(source, unit, precedingCss);
    const customSlugFn = this.options.variableNameSlug;
    if (!customSlugFn) {
      return undefined;
    }

    return typeof customSlugFn === 'function'
      ? customSlugFn(context)
      : buildSlug(customSlugFn, context);
  }

  protected getProps(): IProps {
    const propsObj: IProps = {
      name: this.displayName,
      class: this.className,
      propsAsIs:
        typeof this.component !== 'string' || !allTagsSet.has(this.component),
    };

    // If we found any interpolations, also pass them, so they can be applied
    if (this.interpolations.length) {
      propsObj.vars = {};
      this.interpolations.forEach(({ id, unit, node }) => {
        const items: Expression[] = [this.astService.callExpression(node, [])];

        if (unit) {
          items.push(this.astService.stringLiteral(unit));
        }

        propsObj.vars![id] = items;
      });
    }

    return propsObj;
  }

  protected getTagComponentProps(props: IProps): ObjectExpression {
    const t = this.astService;

    const propExpressions = Object.entries(props)
      .map(([key, value]: [key: string, value: IProps[keyof IProps]]) => {
        if (value === undefined) {
          return null;
        }

        const keyNode = t.identifier(key);

        if (value === null) {
          return t.objectProperty(keyNode, t.nullLiteral());
        }

        if (typeof value === 'string') {
          return t.objectProperty(keyNode, t.stringLiteral(value));
        }

        if (typeof value === 'boolean') {
          return t.objectProperty(keyNode, t.booleanLiteral(value));
        }

        const vars = Object.entries(value).map(([propName, propValue]) => {
          return t.objectProperty(
            t.stringLiteral(propName),
            t.arrayExpression(propValue)
          );
        });

        return t.objectProperty(keyNode, t.objectExpression(vars));
      })
      .filter(isNotNull);

    return t.objectExpression(propExpressions);
  }

  protected getVariableContext(
    source: string,
    unit: string,
    precedingCss: string
  ): IVariableContext {
    const getIndex = () => {
      // eslint-disable-next-line no-plusplus
      return this.#variableIdx++;
    };

    return {
      componentName: this.displayName,
      componentSlug: this.slug,
      get index() {
        return getIndex();
      },
      precedingCss,
      processor: this.constructor.name,
      source,
      unit,
      valueSlug: slugify(source + unit),
    };
  }

  protected getVariableId(
    source: string,
    unit: string,
    precedingCss: string
  ): string {
    const value = source + unit;
    if (!this.#variablesCache.has(value)) {
      const id = this.getCustomVariableId(source, unit, precedingCss);
      if (id) {
        return toValidCSSIdentifier(id);
      }

      const context = this.getVariableContext(source, unit, precedingCss);

      // make the variable unique to this styled component
      this.#variablesCache.set(value, `${this.slug}-${context.index}`);
    }

    return this.#variablesCache.get(value)!;
  }
}
