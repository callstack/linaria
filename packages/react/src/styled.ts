/**
 * This file contains an runtime version of `styled` component. Responsibilities of the component are:
 * - returns ReactElement based on HTML tag used with `styled` or custom React Component
 * - injects classNames for the returned component
 * - injects CSS variables used to define dynamic styles based on props
 */
import React from 'react';
import validAttr from '@emotion/is-prop-valid';
import { cx } from '@linaria/core';
import type { CSSProperties, StyledMeta } from '@linaria/core';

export type NoInfer<A extends any> = [A][A extends any ? 0 : never];

type Options = {
  name: string;
  class: string;
  vars?: {
    [key: string]: [
      string | number | ((props: unknown) => string | number),
      string | void
    ];
  };
};

// Workaround for rest operator
const restOp = (
  obj: Record<string, unknown>,
  keysToExclude: string[]
): Record<string, unknown> =>
  Object.keys(obj)
    .filter((prop) => keysToExclude.indexOf(prop) === -1)
    .reduce((acc, curr) => {
      acc[curr] = obj[curr];
      return acc;
    }, {} as Record<string, unknown>); // rest operator workaround

const warnIfInvalid = (value: any, componentName: string) => {
  if (process.env.NODE_ENV !== 'production') {
    if (
      typeof value === 'string' ||
      // eslint-disable-next-line no-self-compare
      (typeof value === 'number' && isFinite(value))
    ) {
      return;
    }

    const stringified =
      typeof value === 'object' ? JSON.stringify(value) : String(value);

    // eslint-disable-next-line no-console
    console.warn(
      `An interpolation evaluated to '${stringified}' in the component '${componentName}', which is probably a mistake. You should explicitly cast or transform the value to a string.`
    );
  }
};

interface IProps {
  className?: string;
  style?: Record<string, string>;
  [props: string]: unknown;
}

// If styled wraps custom component, that component should have className property
function styled<TConstructor extends React.FunctionComponent<any>>(
  tag: TConstructor extends React.FunctionComponent<infer T>
    ? T extends { className?: string | undefined }
      ? TConstructor
      : never
    : never
): ComponentStyledTag<TConstructor>;
function styled<T>(
  tag: T extends { className?: string | undefined }
    ? React.ComponentType<T>
    : never
): ComponentStyledTag<T>;
function styled<TName extends keyof JSX.IntrinsicElements>(
  tag: TName
): HtmlStyledTag<TName>;
function styled(tag: any): any {
  return (options: Options) => {
    if (process.env.NODE_ENV !== 'production') {
      if (Array.isArray(options)) {
        // We received a strings array since it's used as a tag
        throw new Error(
          'Using the "styled" tag in runtime is not supported. Make sure you have set up the Babel plugin correctly. See https://github.com/callstack/linaria#setup'
        );
      }
    }

    const render = (props: any, ref: any) => {
      const { as: component = tag, class: className } = props;
      const rest = restOp(props, ['as', 'class']);
      let filteredProps: IProps;

      // Check if it's an HTML tag and not a custom element
      if (typeof component === 'string' && component.indexOf('-') === -1) {
        filteredProps = {} as { [key: string]: any };

        // eslint-disable-next-line guard-for-in
        for (const key in rest) {
          if (key === 'as' || validAttr(key)) {
            // Don't pass through invalid attributes to HTML elements
            filteredProps[key] = rest[key];
          }
        }
      } else {
        filteredProps = rest;
      }

      filteredProps.ref = ref;
      filteredProps.className = cx(
        filteredProps.className || className,
        options.class
      );

      const { vars } = options;

      if (vars) {
        const style: { [key: string]: string } = {};

        // eslint-disable-next-line guard-for-in
        for (const name in vars) {
          const variable = vars[name];
          const result = variable[0];
          const unit = variable[1] || '';
          const value = typeof result === 'function' ? result(props) : result;

          warnIfInvalid(value, options.name);

          style[`--${name}`] = `${value}${unit}`;
        }

        const ownStyle = filteredProps.style || {};
        const keys = Object.keys(ownStyle);
        if (keys.length > 0) {
          keys.forEach((key) => {
            style[key] = ownStyle[key];
          });
        }

        filteredProps.style = style;
      }

      if ((tag as any).__linaria && tag !== component) {
        // If the underlying tag is a styled component, forward the `as` prop
        // Otherwise the styles from the underlying component will be ignored
        filteredProps.as = component;

        return React.createElement(tag, filteredProps);
      }
      return React.createElement(component, filteredProps);
    };

    const Result = React.forwardRef
      ? React.forwardRef(render)
      : // React.forwardRef won't available on older React versions and in Preact
        // Fallback to a innerRef prop in that case
        (props: any) => {
          const rest = restOp(props, ['innerRef']);
          return render(rest, props.innerRef);
        };

    (Result as any).displayName = options.name;

    // These properties will be read by the babel plugin for interpolation
    (Result as any).__linaria = {
      className: options.class,
      extends: tag,
    };

    return Result;
  };
}

type StyledComponent<T> = StyledMeta &
  (T extends React.FunctionComponent<any>
    ? T
    : React.FunctionComponent<T & { as?: React.ElementType }>);

type StaticPlaceholder = string | number | CSSProperties | StyledMeta;

type HtmlStyledTag<TName extends keyof JSX.IntrinsicElements> = <
  TAdditionalProps = {}
>(
  strings: TemplateStringsArray,
  ...exprs: Array<
    | StaticPlaceholder
    | ((
        // Without Omit here TS tries to infer TAdditionalProps
        // from a component passed for interpolation
        props: JSX.IntrinsicElements[TName] & Omit<TAdditionalProps, never>
      ) => string | number)
  >
) => StyledComponent<JSX.IntrinsicElements[TName] & TAdditionalProps>;

type ComponentStyledTag<T> = <
  OwnProps = {},
  TrgProps = T extends React.FunctionComponent<infer TProps> ? TProps : T
>(
  strings: TemplateStringsArray,
  // Expressions can contain functions only if wrapped component has style property
  ...exprs: TrgProps extends { style?: React.CSSProperties | undefined }
    ? Array<
        | StaticPlaceholder
        | ((props: NoInfer<OwnProps & TrgProps>) => string | number)
      >
    : StaticPlaceholder[]
) => keyof OwnProps extends never
  ? T extends React.FunctionComponent<any>
    ? StyledMeta & T
    : StyledComponent<TrgProps>
  : StyledComponent<OwnProps & TrgProps>;

type StyledJSXIntrinsics = {
  readonly [P in keyof JSX.IntrinsicElements]: HtmlStyledTag<P>;
};

export type Styled = typeof styled & StyledJSXIntrinsics;

export default (process.env.NODE_ENV !== 'production'
  ? new Proxy(styled, {
      get(o, prop: keyof JSX.IntrinsicElements) {
        return o(prop);
      },
    })
  : styled) as Styled;
