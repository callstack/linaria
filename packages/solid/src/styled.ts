import type { JSX, Component } from 'solid-js';

import type { CSSProperties } from '@linaria/core';
import type { StyledMeta } from '@linaria/tags';

type TagExpression<Props> =
  | string
  | number
  | CSSProperties
  | StyledMeta
  | ((props: Props) => string | number | undefined);

export interface StyledComponent<Props> extends StyledMeta, Component<Props> {}

export interface StyledTag<Props> {
  (
    strings: TemplateStringsArray,
    ...expressions: readonly TagExpression<Props>[]
  ): StyledComponent<Props>;
  <AdditionalProps = Record<string, unknown>>(
    strings: TemplateStringsArray,
    ...expressions: readonly TagExpression<Props & AdditionalProps>[]
  ): StyledComponent<Props & AdditionalProps>;
}

export interface Styled {
  <Props>(component: Component<Props>): StyledTag<Props>;
  <TagName extends keyof JSX.IntrinsicElements>(tagName: TagName): StyledTag<
    JSX.IntrinsicElements[TagName]
  >;
  (component: 'The target component must have a className prop'): never;
}

type StyledJSXIntrinsics = {
  readonly [TagName in keyof JSX.IntrinsicElements]: StyledTag<
    JSX.IntrinsicElements[TagName]
  >;
};

export const styled: Styled & StyledJSXIntrinsics = (() => {
  throw new Error(
    'Using the "styled" tag in runtime is not supported. Make sure you have set up the Babel plugin correctly. See https://github.com/callstack/linaria#setup'
  );
}) as never;
