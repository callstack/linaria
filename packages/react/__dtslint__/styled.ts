/* tslint:disable:no-unnecessary-generics */
// eslint-disable-next-line import/no-extraneous-dependencies
import * as React from 'react';
import { css } from '@linaria/core';
import { styled } from '../src';

// eslint-disable-next-line @typescript-eslint/no-unused-vars
function isExtends<C, T>(arg1?: C, arg2?: T): C extends T ? 'extends' : never {
  // It will never be executed, so the result doesn't matter.
  return null as any;
}

const Fabric =
  <T>(): React.FC<T> =>
  (props) =>
    React.createElement('div', props);

const Header = (p: { children: string }) => React.createElement('h1', p);

const Generic = <T>(p: T & { className?: string }) =>
  React.createElement('h1', p);

const StyledDiv = styled.div``;
// $ExpectType "extends"
isExtends<typeof StyledDiv, React.FC<React.DetailedHTMLProps<any, any>>>();

// component should have className property
// $ExpectError
styled(Fabric<{ a: string }>())``;

// className property should be string
// $ExpectError
styled(Fabric<{ className: number }>())``;

const SimplestComponent = styled(Fabric<{ className: string }>())``;
// $ExpectType "extends"
isExtends<typeof SimplestComponent, React.FC<{ className: string }>>();

styled(Fabric<{ className: string }>())`
  // component should have style property
  // $ExpectError
  color: ${() => 'red'};
`;

styled(Fabric<{ className: string }>())`
  // it looks like function, but it's a reference to another styled component
  & > ${SimplestComponent} {
    color: red;
  }
`;

styled(Fabric<{ className: string }>())`
  // it looks like the previous test, but it references a non-linaria component
  // $ExpectError
  & > ${Header} {
    color: red;
  }
`;

styled(Fabric<{ className: string; style: {} }>())`
  color: ${() => 'red'};
`;

styled(Fabric<{ className: string; style: {} }>())`
  // color should be defined in props
  // $ExpectError
  color: ${(props) => props.color};
`;

styled(Fabric<{ className: string; style: {}; color: 'red' | 'blue' }>())`
  & > ${SimplestComponent} {
    color: ${(props) => props.color};
  }
`;

// $ExpectType number
Generic({ children: 123 }).props.children;

const StyledGeneric = styled(Generic)``;
// $ExpectType number
StyledGeneric({ children: 123 }).props.children;

styled.a`
  & > ${SimplestComponent} {
    color: red;
  }
`({ href: 'about:blank' });

((/* Issue #536 */) => {
  const Title = styled.div<{ background: string }>`
    background: ${(props) => props.background};
  `;

  // $ExpectType "extends"
  isExtends<typeof Title, React.FC<{ background: string }>>();

  css`
    ${Title} {
      color: green;
    }
  `;
})();

((/* Issue #622 */) => {
  const Wrapper = styled.div<{ prop1: boolean }>`
    width: 1em;
    background-color: ${(props) => (props.prop1 ? 'transparent' : 'green')};
  `;

  const Custom: React.FC<{ className?: string; id: number }> = () => null;

  const tag = styled(Custom);
  const Card = tag`
    ${Wrapper} {
      color: green;
    }
  `;

  // $ExpectType Validator<number> | undefined
  Card.propTypes!.id;

  const styledTag = styled(Wrapper);

  const NewWrapper = styledTag<{ prop2: string }>`
    width: 2em;
    background-color: ${(props) => (props.prop1 ? 'transparent' : 'red')};
    color: ${(props) => props.prop2};
  `;

  // $ExpectType Validator<boolean> | undefined
  NewWrapper.propTypes!.prop1;

  // $ExpectType Validator<string> | undefined
  NewWrapper.propTypes!.prop2;
})();

((/* Issue #844 */) => {
  type GridProps = { container?: false } | { container: true; spacing: number };

  const Grid: React.FC<GridProps & { className?: string }> = () => null;

  // Type 'false' is not assignable to type 'true'
  // $ExpectError
  React.createElement(Grid, { container: false, spacing: 8 });

  React.createElement(Grid, { container: true, spacing: 8 });

  styled(Grid)``;
})();

((/* Issue #872 */) => {
  interface BaseProps {
    className?: string;
    style?: React.CSSProperties;
  }

  interface ComponentProps extends BaseProps {
    title: string;
  }

  const Flow = <TProps extends BaseProps>(Cmp: React.FC<TProps>) =>
    styled(Cmp)`
      display: flow;
    `;

  const Component: React.FC<ComponentProps> = (props) =>
    React.createElement('div', props);

  const Implementation = Flow(Component);

  (() => React.createElement(Implementation, { title: 'Title' }))();
})();
