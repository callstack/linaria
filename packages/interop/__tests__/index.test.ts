const babel = require('@babel/core');
const dedent = require('dedent');

const plugin = require('../src');

const getCode = (src: string) => {
  const { code } = babel.transform(dedent(src), {
    filename: 'test.js',
    babelrc: false,
    plugins: [plugin],
  });
  return code;
};

describe('styled-components', () => {
  it('keeps linaria', () => {
    const code = getCode(`
    import styled from "linaria/react";
    import Title from "./Title";

    export default styled.h1\`
      & > ${'${Title}'} {
        color: red;
      }
    \`;
  `);

    expect(code).toMatchSnapshot();
  });

  it('css', () => {
    const code = getCode(`
    import { css } from "styled-components";
    import Title from "./Title";

    export default css\`
      & > ${'${Title}'} {
        color: red;
      }
    \`;
  `);

    expect(code).toMatchSnapshot();
  });

  it('styled.h1', () => {
    const code = getCode(`
    import styled from "styled-components";
    import Title from "./Title";

    export default styled.h1\`
      & > ${'${Title}'} {
        color: red;
      }
    \`;
  `);

    expect(code).toMatchSnapshot();
  });

  it('member expression as selector', () => {
    const code = getCode(`
    import styled from "styled-components";
    import Title from "./Title";

    export default styled.h1\`
      & > ${'${Title.Small}'} {
        color: red;
      }
    \`;
  `);

    expect(code).toMatchSnapshot();
  });

  it('styled(Cmp)', () => {
    const code = getCode(`
    import styled from "styled-components";
    import Cmp from "./Cmp";
    import Title from "./Title";

    export default styled(Cmp)\`
      & > ${'${Title}'} {
        color: red;
      }
    \`;
  `);

    expect(code).toMatchSnapshot();
  });

  it('styled(Cmp).attrs({})', () => {
    const code = getCode(`
    import styled from "styled-components";
    import Cmp from "./Cmp";
    import Title from "./Title";

    export default styled(Cmp).attrs(() => ({}))\`
      & > ${'${Title}'} {
        color: red;
      }
    \`;
  `);

    expect(code).toMatchSnapshot();
  });

  it('styled.h1.attrs({})', () => {
    const code = getCode(`
    import styled from "styled-components";
    import Title from "./Title";

    export default styled.h1.attrs(() => ({}))\`
      & > ${'${Title}'} {
        color: red;
      }
    \`;
  `);

    expect(code).toMatchSnapshot();
  });
});
