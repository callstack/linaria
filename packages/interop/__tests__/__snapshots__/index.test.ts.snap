// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styled-components css 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = void 0;
var _styledComponents = require(\\"styled-components\\");
var _Title = _interopRequireDefault(require(\\"./Title\\"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var _default = (0, _styledComponents.css)\`
  & > \${(i => i && i.__linaria ? '.' + i.__linaria.className : i)(_Title.default)} {
    color: red;
  }
\`;
exports.default = _default;"
`;

exports[`styled-components keeps linaria 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require(\\"linaria/react\\"));
var _Title = _interopRequireDefault(require(\\"./Title\\"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var _default = _react.default.h1\`
  & > \${_Title.default} {
    color: red;
  }
\`;
exports.default = _default;"
`;

exports[`styled-components member expression as selector 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = void 0;
var _styledComponents = _interopRequireDefault(require(\\"styled-components\\"));
var _Title = _interopRequireDefault(require(\\"./Title\\"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var _default = _styledComponents.default.h1\`
  & > \${(i => i && i.__linaria ? '.' + i.__linaria.className : i)(_Title.default.Small)} {
    color: red;
  }
\`;
exports.default = _default;"
`;

exports[`styled-components styled(Cmp) 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = void 0;
var _styledComponents = _interopRequireDefault(require(\\"styled-components\\"));
var _Cmp = _interopRequireDefault(require(\\"./Cmp\\"));
var _Title = _interopRequireDefault(require(\\"./Title\\"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var _default = (0, _styledComponents.default)(_Cmp.default)\`
  & > \${(i => i && i.__linaria ? '.' + i.__linaria.className : i)(_Title.default)} {
    color: red;
  }
\`;
exports.default = _default;"
`;

exports[`styled-components styled(Cmp).attrs({}) 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = void 0;
var _styledComponents = _interopRequireDefault(require(\\"styled-components\\"));
var _Cmp = _interopRequireDefault(require(\\"./Cmp\\"));
var _Title = _interopRequireDefault(require(\\"./Title\\"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var _default = (0, _styledComponents.default)(_Cmp.default).attrs(() => ({}))\`
  & > \${(i => i && i.__linaria ? '.' + i.__linaria.className : i)(_Title.default)} {
    color: red;
  }
\`;
exports.default = _default;"
`;

exports[`styled-components styled.h1 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = void 0;
var _styledComponents = _interopRequireDefault(require(\\"styled-components\\"));
var _Title = _interopRequireDefault(require(\\"./Title\\"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var _default = _styledComponents.default.h1\`
  & > \${(i => i && i.__linaria ? '.' + i.__linaria.className : i)(_Title.default)} {
    color: red;
  }
\`;
exports.default = _default;"
`;

exports[`styled-components styled.h1.attrs({}) 1`] = `
"\\"use strict\\";

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.default = void 0;
var _styledComponents = _interopRequireDefault(require(\\"styled-components\\"));
var _Title = _interopRequireDefault(require(\\"./Title\\"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
var _default = _styledComponents.default.h1.attrs(() => ({}))\`
  & > \${(i => i && i.__linaria ? '.' + i.__linaria.className : i)(_Title.default)} {
    color: red;
  }
\`;
exports.default = _default;"
`;
